### PR TITLE
add module pacman_key

### PR DESCRIPTION
**Pull request by grawlinson**
##### SUMMARY

Manage [pacman](https://wiki.archlinux.org/index.php/Pacman)'s keyring of trusted keys with [pacman-key](https://wiki.archlinux.org/index.php/Pacman/Package_signing). Similar to the [apt_key](https://docs.ansible.com/ansible/latest/modules/apt_key_module.html) & [rpm_key](https://docs.ansible.com/ansible/latest/modules/rpm_key_module.html) modules. Useful when adding additional repositories that have signed packages/databases.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
pacman_key

EDIT: This is originally from this [PR](https://github.com/ansible/ansible/pull/68232) on ansible/ansible.

### CODE REVIEW

File: plugins/modules/packaging/os/pacman_key.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:2]: You need to add a symlink from `plugins/modules/` to this file.

File: plugins/modules/packaging/os/pacman_key.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: pacman_key
+author:
+- George Rawlinson (@grawlinson) <george@rawlinson.net.nz>
+version_added: "2.10"
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:19]:
```suggestion
version_added: "1.1.0"
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+short_description: Manage pacman's list of trusted keys
+description:
+- Add or remove gpg keys from the pacman keyring.
+notes:
+- Use full key id (16 characters) and fingerprint (40 characters) to avoid key collisions.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:24]:
```suggestion
- Use full key ID (16 characters) and fingerprint (40 characters) to avoid key collisions.
```
(same below)

File: plugins/modules/packaging/os/pacman_key.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+    data:
+        description:
+            - The keyfile contents to add to the keyring.
+            - Must be of "PGP PUBLIC KEY BLOCK" type.
+        type: str
+    file:
+        description:
+            - The path to a keyfile on the remote server to add to the keyring.
+            - Remote file should be of "PGP PUBLIC KEY BLOCK" type.
+        type: path
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:48]: I guess `data`, `file` and `url` are mutually exclusive? If yes, that should be mentioned in the documentation.

File: plugins/modules/packaging/os/pacman_key.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+            elif url:
+                data = self.fetch_key(url)
+                file = self.save_key(data)
+                self.add_key(file, keyid, fingerprint, keyring)
+                module.exit_json(changed=True)
+            elif keyserver:
+                self.recv_key(keyid, keyserver, keyring)
+                module.exit_json(changed=True)
+            else:
+                module.fail_json(msg="expected one of: data, file, url, keyserver. got none")
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:175]: Check this earlier. check_mode should also fail when this is not the case, and the module should also fail if this is not the case and they key already exists in the keyring.

**Comment by grawlinson** [comment]
@felixfontein Thanks for the exceptional feedback! I'll implement your suggestions when I have the time.

File: plugins/modules/packaging/os/pacman_key.py [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+version_added: "1.1.0"
+short_description: Manage pacman's list of trusted keys
+description:
+- Add or remove gpg keys from the pacman keyring.
+notes:
+- Use full key ID (16 characters) and fingerprint (40 characters) to avoid key collisions.
+- If you specify both the key id and the URL with C(state=present), the task can verify or add the key as needed.
+- By default, keys will be locally signed after being imported into the keyring.
+- If the specified key id exists in the keyring, the key will not be added.
+- data, file, and url are mutually exclusive.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:28]:
```suggestion
- I(data), I(file), and I(url) are mutually exclusive.
```
That should better be mentioned in the description for all of these options. People tend to ignore most of the documentation anyway, and the chance of them noticing this part there is a lot better :)

File: plugins/modules/packaging/os/pacman_key.py (hunk 5 above)
Comment (grawlinson) [reply plugins/modules/packaging/os/pacman_key.py]: Latest commit _should_ fix this. It's a bit early in the morning to be coding, do you mind checking this over?

Comment (felixfontein) [reply plugins/modules/packaging/os/pacman_key.py]: Looks good to me!

Comment (grawlinson) [reply plugins/modules/packaging/os/pacman_key.py]: Fantastic. I’ll throw together some unit tests. Can you recommend any examples in pre-existing modules or documentation that I should browse?

Comment (felixfontein) [reply plugins/modules/packaging/os/pacman_key.py]: Not in particular. The tests for the ufw module look helpful: https://github.com/ansible-collections/community.general/blob/main/tests/unit/plugins/modules/system/test_ufw.py The ufw module is mainly executing the `ufw` binary and reacts on its output, so you can get some ideas from there.

File: plugins/modules/packaging/os/pacman_key.py [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+
+# Copyright: (c) 2019, George Rawlinson <george@rawlinson.net.nz>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:12]:
```suggestion
```
No needed any more

File: plugins/modules/packaging/os/pacman_key.py [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+
+DOCUMENTATION = '''
+---
+module: pacman_key
+author:
+- George Rawlinson (@grawlinson) <george@rawlinson.net.nz>
+version_added: "1.1.0"
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:19]:
```suggestion
version_added: "1.3.0"
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: pacman_key
+author:
+- George Rawlinson (@grawlinson) <george@rawlinson.net.nz>
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:18]:
```suggestion
- George Rawlinson (@grawlinson)
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+        description:
+            - Ensures that the key is present (added) or absent (revoked).
+        default: present
+        choices: [ absent, present ]
+        type: str
+'''
+
+EXAMPLES = '''
+- name: Import a key via local file
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:85]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+    data: "{{ lookup('file', 'keyfile.asc') }}"
+    state: present
+
+- name: Import a key via remote file
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:90]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+    file: /tmp/keyfile.asc
+    state: present
+
+- name: Import a key via url
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:95]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    url: https://domain.tld/keys/keyfile.asc
+    state: present
+
+- name: Import a key via keyserver
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:101]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    keyserver: keyserver.domain.tld
+
+- name: Import a key into an alternative keyring
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:106]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    id: 01234567890ABCDE01234567890ABCDE12345678
+    file: /tmp/keyfile.asc
+    keyring: /etc/pacman.d/gnupg-alternative
+
+- name: Remove a key from the keyring
+  pacman_key:
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:112]:
```suggestion
  community.general.pacman_key:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+version_added: "1.1.0"
+short_description: Manage pacman's list of trusted keys
+description:
+- Add or remove gpg keys from the pacman keyring.
+notes:
+- Use full key ID (16 characters) and fingerprint (40 characters) to avoid key collisions.
+- If you specify both the key id and the URL with C(state=present), the task can verify or add the key as needed.
+- By default, keys will be locally signed after being imported into the keyring.
+- If the specified key id exists in the keyring, the key will not be added.
+- I(data), I(file), and I(url) are mutually exclusive.
```
Comment (Andersson007) [inline plugins/modules/packaging/os/pacman_key.py:28]:
```suggestion
- I(data), I(file), and I(url) are mutually exclusive.
- Supports C(check_mode).
```

**Review by Andersson007 — COMMENTED** [review]
@grawlinson thanks for the new module!
To have unit or integration tests would be good

**Comment by felixfontein** [comment]
@grawlinson are you still working on this?

needs_info

**Comment by grawlinson** [comment]
Since the last few messages, I've got a new laptop so this hasn't been a pressing priority.

I'll hop on IRC and ask a few questions in the next week or so.

**Comment by russoz** [comment]
@grawlinson any updates on this?

**Comment by russoz** [comment]
-label needs_triage

**Comment by felixfontein** [comment]
@grawlinson ping

needs_info

**Comment by grawlinson** [comment]
The hoops I have to jump through just to add unit tests are a major pain point, as the documentation isn't all in one place.

**Comment by grawlinson** [comment]
Running `ansible-test units --docker default -v` results in Docker saying that the Linux kernel crashed and this error:

`failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/027c68e490f797c1fb61f605688b349e3fed3904a38fb388f3246a736717335f/merged: too many levels of symbolic links`

**Comment by grawlinson** [comment]
I'm not interested in working around incomplete documentation or broken tooling. Can this just be merged as-is? I've been using it in production since I developed it.

**Comment by felixfontein** [comment]
> Can this just be merged as-is? I've been using it in production since I developed it.

Sorry, no. We no longer merge new plugins/modules without test. (We did that in the past, and almost every time it was a huge mistake.)

**Comment by felixfontein** [comment]
> Running `ansible-test units --docker default -v` results in Docker saying that the Linux kernel crashed and this error:
> 
> `failed to register layer: error creating overlay mount to /var/lib/docker/overlay2/027c68e490f797c1fb61f605688b349e3fed3904a38fb388f3246a736717335f/merged: too many levels of symbolic links`

Sonds a bit like your docker setup is broken. Sometimes in the past, I had to reboot my machine to work around such errors. Maybe that helps.

**Review by russoz — CHANGES_REQUESTED** [review]

File: plugins/modules/packaging/os/pacman_key.py [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+            url=dict(type='str'),
+            keyserver=dict(type='str'),
+            fingerprint=dict(type='str'),
+            force_update=dict(type='bool', default=False),
+            keyring=dict(type='path'),
+            state=dict(type='str', default='present', choices=['absent', 'present']),
+        ),
+        supports_check_mode=True,
+        mutually_exclusive=(('data', 'file', 'url', 'keyserver'),),
+    )
```
[the remark this answers is not in the document]
Comment (grawlinson) [reply plugins/modules/packaging/os/pacman_key.py:334]: Aaaahhh, where were you a year ago? :P

File: plugins/modules/packaging/os/pacman_key.py [hunk 18]
```diff
@@ -331,6 +322,7 @@ def main():
         ),
         supports_check_mode=True,
         mutually_exclusive=(('data', 'file', 'url', 'keyserver'),),
+        required_one_of=[('data', 'file', 'url', 'keyserver'),],
```
Comment (russoz) [inline plugins/modules/packaging/os/pacman_key.py:325]: The previous code requires one of those **only** when `state` is `present`. This code imposes that on all states.

Comment (russoz) [reply plugins/modules/packaging/os/pacman_key.py:325]: Yes, I just realized it was my own comment :)

File: plugins/modules/packaging/os/pacman_key.py [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+    def add_key(self, keyfile, keyid, fingerprint=None, keyring=None):
+        """Add key to pacman's keyring"""
+        cmd = [self.pacman_key]
+        if keyring is not None:
+            cmd.extend(['--gpgdir', keyring])
+
+        if fingerprint is not None:
+            self.verify_keyfile(keyfile, keyid, fingerprint)
+
+        rc, stdout, stderr = self.module.run_command(cmd + ['--add', keyfile])
```
[the remark this answers is not in the document]
Comment (grawlinson) [reply plugins/modules/packaging/os/pacman_key.py:252]: Yeah, I'll probably do that except for the `key_in_keyring()` function

**Comment by russoz** [comment]
@grawlinson you will also need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to). And this collection no longer accepts new modules without tests.

As of the `required_if` parameter, I know I was the one suggesting that, but on second thought it might be a bad idea, for state=absent.

**Comment by felixfontein** [comment]
> @grawlinson you will also need a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

New modules must not have changelog fragments.

**Comment by yan12125** [comment]
The examples seem incorrect. Each `id` should be 16-characters long, while in examples each `id` is 40-characters long. By the way, how about using full fingerprints (40-characters) everywhere? That brings less confusion and slightly improves security (even harder for collisions).

**Comment by grawlinson** [comment]
> The examples seem incorrect. Each `id` should be 16-characters long, while in examples each `id` is 40-characters long. By the way, how about using full fingerprints (40-characters) everywhere? That brings less confusion and slightly improves security (even harder for collisions).

Sounds good to me, I'll do that soon.

**Comment by grawlinson** [comment]
It is currently taking me about 30 minutes just to download the docker images required for `ansible-test`, let alone the various network calls that each invocation of `ansible-test` does. I've got about 10 tabs open just looking at how Ansible does its tests and I'm not exactly getting any closer to doing any of these tests. I am not exactly impressed with any of this.

Off the top of my head, all I need to do is stub out `module.get_bin_path()` for `gpg` and `pacman_key`, about 6 invocations of `module.run_command()` in various orders. From looking at other tests, the boilerplate code required to stub all this stuff out is looking like a nightmare.

**Comment by yan12125** [comment]
> Sounds good to me, I'll do that soon.

Thanks for the update! Here are just some minor typos. With these changes everything works fine :)
```Diff
diff --git a/plugins/modules/packaging/os/pacman_key.py b/plugins/modules/packaging/os/pacman_key.py
index 0583f612..83abb9f6 100644
--- a/plugins/modules/packaging/os/pacman_key.py
+++ b/plugins/modules/packaging/os/pacman_key.py
@@ -139,7 +139,7 @@ class PacmanKey(object):
         keyid = self.sanitise_identifier(keyid)
         if fingerprint:
             fingerprint = self.sanitise_identifier(fingerprint)
-        key_present = self.key_in_keyring(keyid, keyring)
+        key_present = self.key_in_keyring(keyring, keyid)
 
         if (
             state == "present"
@@ -218,7 +218,7 @@ class PacmanKey(object):
         if keyserver is not None:
             cmd.extend(['--keyserver', keyserver])
         self.module.run_command(cmd + ['--recv-keys', keyid], check_rc=True)
-        self.lsign_key(keyid, keyring)
+        self.lsign_key(keyring, keyid)
 
     def lsign_key(self, keyring, keyid):
         """Locally sign key"""
@@ -240,7 +240,7 @@ class PacmanKey(object):
         if fingerprint is not None:
             self.verify_keyfile(keyfile, keyid, fingerprint)
         self.module.run_command(cmd + ['--add', keyfile], check_rc=True)
-        self.lsign_key(keyid, keyring)
+        self.lsign_key(keyring, keyid)
 
     def remove_key(self, keyring, keyid):
         """Remove key from pacman's keyring"""
```

> It is currently taking me about 30 minutes just to download the docker images required for `ansible-test`, let alone the various network calls that each invocation of `ansible-test` does. ...

FWIW, when I was working on https://github.com/ansible-collections/community.general/pull/2635, I found that running ansible-test without docker seems working, too. For example,

```
ansible-test units --python 3.9 -v tests/unit/plugins/modules/net_tools/test_nmcli.py
```

> Off the top of my head, all I need to do is stub out `module.get_bin_path()` for `gpg` and `pacman_key`, about 6 invocations of `module.run_command()` in various orders. ...

If I understand `test_nmcli.py` correctly, its `mocker_set` method mocks out `AnsibleModule.get_bin_path()` and `Nmcli.execute_command()` (which calls `run_command()`) to return specified return some specified values. Such a pattern might be useful for testing `pacman_key`, too.

https://github.com/ansible-collections/community.general/blob/3.1.0/tests/unit/plugins/modules/net_tools/test_nmcli.py#L377

File: plugins/modules/packaging/os/pacman_key.py [hunk 20]
```diff
@@ -322,7 +331,6 @@ def main():
         ),
         supports_check_mode=True,
         mutually_exclusive=(('data', 'file', 'url', 'keyserver'),),
-        required_one_of=[('data', 'file', 'url', 'keyserver'),],
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py]: Instead of removing this (and adding manual checks), better add
```.py
required_if=[('state', 'present', ('data', 'file', 'url', 'keyserver'), True)],
```
([Docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#dependencies-between-module-options))

File: plugins/modules/packaging/os/pacman_key.py [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+    module = AnsibleModule(
+        argument_spec=dict(
+            id=dict(type='str', required=True),
+            data=dict(type='str'),
+            file=dict(type='path'),
+            url=dict(type='str'),
+            keyserver=dict(type='str'),
+            fingerprint=dict(type='str'),
+            force_update=dict(type='bool', default=False),
+            keyring=dict(type='path', default='/etc/pacman.d/gnupg'),
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:298]:
```suggestion
            keyring=dict(type='path'),
```
Alternatively, add the default in the documentation as well.

File: plugins/modules/packaging/os/pacman_key.py [hunk 22]
```diff
@@ -0,0 +1,308 @@
...
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: pacman_key
+author:
+- George Rawlinson (@grawlinson)
+version_added: "1.3.0"
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:15]:
```suggestion
version_added: "3.2.0"
```

**Comment by grawlinson** [comment]
Some initial tests up, please don't merge until I've added some better ones.

**Comment by grawlinson** [comment]
Still have at least 4 more tests planned, plus some edge-case checks, so this still isn't ready to merge.

If anything in the current tests can be improved, I'd love to know.

File: tests/unit/plugins/modules/packaging/os/test_pacman_key.py [hunk 23]
```diff
@@ -0,0 +1,444 @@
...
+    with pytest.raises(SystemExit):
+        pacman_key.main()
+
+    # capture std{out,err}
+    out, err = capfd.readouterr()
+    results = json.loads(out)
+    print(results)
+
+    # assertion time!
+    assert 'failed' in results
```
Comment (felixfontein) [inline tests/unit/plugins/modules/packaging/os/test_pacman_key.py:103]: Why not
```suggestion
    assert results['failed'] is True
```
? A `failed: false` in the results is something you don't want to accept :)

File: tests/unit/plugins/modules/packaging/os/test_pacman_key.py [hunk 24]
```diff
@@ -0,0 +1,444 @@
...
+    if 'msg' in results:
```
Comment (felixfontein) [inline tests/unit/plugins/modules/packaging/os/test_pacman_key.py:104]:
```suggestion
    if 'msg' in expected:
```

File: tests/unit/plugins/modules/packaging/os/test_pacman_key.py [hunk 25]
```diff
@@ -0,0 +1,444 @@
...
+        pacman_key.main()
+
+    # capture std{out,err}
+    out, err = capfd.readouterr()
+    results = json.loads(out)
+
+    # assertion time!
+    assert 'changed' in results
+    assert results['changed'] == testcase['changed']
+    if 'msg' in results:
```
Comment (felixfontein) [inline tests/unit/plugins/modules/packaging/os/test_pacman_key.py:437]:
```suggestion
    if 'msg' in expected:
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 26]
```diff
@@ -0,0 +1,320 @@
...
+        type: str
+    force_update:
+        description:
+            - This forces the key to be updated if it already exists in the keyring.
+        type: bool
+        default: no
+    keyring:
+        description:
+            - The full path to the keyring folder on the remote server.
+            - If not specified, module will use pacman's default (/etc/pacman.d/gnupg).
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:69]:
```suggestion
            - If not specified, module will use pacman's default (C(/etc/pacman.d/gnupg)).
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 27]
```diff
@@ -0,0 +1,320 @@
...
+        type: str
+    file:
+        description:
+            - The path to a keyfile on the remote server to add to the keyring.
+            - Remote file should be of "PGP PUBLIC KEY BLOCK" type.
+        type: path
+    url:
+        description:
+            - The URL to retrieve keyfile from.
+            - Remote file should be of "PGP PUBLIC KEY BLOCK" type.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:50]:
```suggestion
            - Remote file should be of C(PGP PUBLIC KEY BLOCK) type.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 28]
```diff
@@ -0,0 +1,320 @@
...
+        type: str
+    data:
+        description:
+            - The keyfile contents to add to the keyring.
+            - Must be of "PGP PUBLIC KEY BLOCK" type.
+        type: str
+    file:
+        description:
+            - The path to a keyfile on the remote server to add to the keyring.
+            - Remote file should be of "PGP PUBLIC KEY BLOCK" type.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:45]:
```suggestion
            - Remote file should be of C(PGP PUBLIC KEY BLOCK) type.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 29]
```diff
@@ -0,0 +1,320 @@
...
+        description:
+            - The 40 character identifier of the key.
+            - Including this allows check mode to correctly report the changed state.
+            - Do not specify a subkey id, instead specify the master key id.
+        required: yes
+        type: str
+    data:
+        description:
+            - The keyfile contents to add to the keyring.
+            - Must be of "PGP PUBLIC KEY BLOCK" type.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:40]:
```suggestion
            - Must be of C(PGP PUBLIC KEY BLOCK) type.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 30]
```diff
@@ -0,0 +1,320 @@
...
+- Supports C(check_mode).
+requirements:
+- gpg
+- pacman-key
+options:
+    id:
+        description:
+            - The 40 character identifier of the key.
+            - Including this allows check mode to correctly report the changed state.
+            - Do not specify a subkey id, instead specify the master key id.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:34]:
```suggestion
            - Do not specify a subkey ID, instead specify the master key ID.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 31]
```diff
@@ -0,0 +1,320 @@
...
+module: pacman_key
+author:
+- George Rawlinson (@grawlinson)
+version_added: "3.2.0"
+short_description: Manage pacman's list of trusted keys
+description:
+- Add or remove gpg keys from the pacman keyring.
+notes:
+- Use full-length identifier (40 characters) for key ID and fingerprint to avoid key collisions.
+- If you specify both the key id and the URL with C(state=present), the task can verify or add the key as needed.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:21]:
```suggestion
- If you specify both the key ID and the URL with I(state=present), the task can verify or add the key as needed.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 32]
```diff
@@ -0,0 +1,320 @@
...
+- By default, keys will be locally signed after being imported into the keyring.
+- If the specified key id exists in the keyring, the key will not be added.
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:23]:
```suggestion
- If the specified key ID exists in the keyring, the key will not be added.
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 33]
```diff
@@ -0,0 +1,320 @@
...
+    def key_in_keyring(self, keyring, keyid):
+        "Check if the keyid is in pacman's keyring"
+        rc, stdout, stderr = self.module.run_command(
+            [
+                self.gpg,
+                '--with-colons',
+                '--batch',
+                '--no-tty',
+                '--no-default-keyring',
+                '--keyring=' + keyring + '/pubring.gpg',
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:286]:
```suggestion
                '--keyring=%s/pubring.gpg' % keyring,
```

**Comment by felixfontein** [comment]
Please note that community.general will be released next Tuesday. Would be great if we can get this PR merged for that :)

**Comment by grawlinson** [comment]
I think module test coverage is at ~90% at the moment. Is there a threshold that new modules need?

File: plugins/modules/packaging/os/pacman_key.py [hunk 34]
```diff
@@ -0,0 +1,315 @@
...
+        # sanitise key ID & check if key exists in the keyring
+        keyid = self.sanitise_keyid(keyid)
+        key_present = self.key_in_keyring(keyring, keyid)
+
+        # check mode
+        if module.check_mode:
+            if state == "present":
+                if (key_present and force_update) or not key_present:
+                    module.exit_json(changed=True)
+                module.exit_json(changed=False)
```
Comment (russoz) [inline plugins/modules/packaging/os/pacman_key.py:152]: Nit picking alert!
```suggestion
                changed = (key_present and force_update) or not key_present
                module.exit_json(changed=changed)
```
Yours truly personal preferences lean towards less conditional statements. Feel free to ignore as it is just a matter of styles.

**Comment by russoz** [comment]
> I think module test coverage is at ~90% at the moment. Is there a threshold that new modules need?

Not that I am aware @grawlinson . (Maybe we should, but to establish *any* threshold without adding test coverage to all modules (many are at 0% today) wouldn't feel right, IMO).

**Review by russoz — APPROVED** [review]
Looking good to me! Thanks @grawlinson for your contribution! 😁

File: plugins/modules/packaging/os/pacman_key.py [hunk 35]
```diff
@@ -0,0 +1,314 @@
...
+requirements:
+- gpg
+- pacman-key
+options:
+    id:
+        description:
+            - The 40 character identifier of the key.
+            - Including this allows check mode to correctly report the changed state.
+            - Do not specify a subkey ID, instead specify the primary key ID.
+        required: yes
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:35]: YAML booleans should always be `true`/`false`.
```suggestion
        required: true
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 36]
```diff
@@ -0,0 +1,314 @@
...
+        type: str
+    keyserver:
+        description:
+            - The keyserver used to retrieve key from.
+        type: str
+    verify:
+        description:
+            - Whether or not to verify the keyfile's key ID against specified key ID.
+        type: bool
+        default: yes
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:60]:
```suggestion
        default: true
```

File: plugins/modules/packaging/os/pacman_key.py [hunk 37]
```diff
@@ -0,0 +1,314 @@
...
+    force_update:
+        description:
+            - This forces the key to be updated if it already exists in the keyring.
+        type: bool
+        default: no
```
Comment (felixfontein) [inline plugins/modules/packaging/os/pacman_key.py:65]:
```suggestion
        default: false
```

**Review by felixfontein — COMMENTED** [review]
Something I missed before. Besides that, looks good!

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@grawlinson thanks a lot for your contribution!
@Andersson007 @russoz @yan12125 thanks for reviewing etc.!

**Comment by djmattyg007** [comment]
Wow, nice work and thank you for persevering <3
